clusterman/autoscaler/pool_manager.py

```diff
@@ -375,79 +418,161 @@ def _choose_nodes_to_prune(
 
-        def is_constrained(group):
+        # when coeff is positive, if any resource is under our target, we need to keep scaling.
+        # when coeff is negative, if any resource is under or at our target, we can stop scaling.
+        def keep_scaling() -> bool:
+            total_target_capacity = sum([a.target_capacity for a in actions.values()], ClustermanResources())
             if coeff > 0:
-                return targets[group.id] + coeff > group.max_capacity
+                return total_target_capacity.any_lt(new_target_capacity)
             else:
-                return targets[group.id] + coeff < group.min_capacity
+                return not total_target_capacity.any_le(new_target_capacity)
+
+        def valid_options() -> Iterator[Tuple[str, ClusterNodeMetadata]]:
+            """generate the actions we can take without violating constraints, e.g. launch c5.12xlarge in group A,
+            launch m5.9xlarge in group B."""
+
+            total_target_capacity = sum([a.target_capacity for a in actions.values()], ClustermanResources())
+
+            if coeff > 0:
+                for group in non_stale_groups:
+                    for option in group.scale_up_options():
+                        total_with_option = total_target_capacity + option.agent.total_resources
+                        group_with_option = actions[group.id].target_capacity + option.agent.total_resources
+                        # Disqualify options that would push us beyond our limits.
+                        if group_with_option.all_le(group.max_capacity) and total_with_option.all_le(self.max_capacity):
+                            # Only consider an option if it gets us closer to our target.
+                            if total_with_option.clamp(upper_bound=new_target_capacity).any_gt(total_target_capacity):
+                                yield (group.id, option)
+            else:
+                for option in self.get_node_metadatas():
+                    group_id = option.instance.group_id
+                    group = self.resource_groups[group_id]
+
+                    total_with_option = total_target_capacity - option.agent.total_resources
+                    group_with_option = actions[group.id].target_capacity - option.agent.total_resources
+
+                    # Disqualify options that would push us below minimum or new_target_capacity.
+                    if all([
+                        group_with_option.all_ge(group.min_capacity),
+                        total_with_option.all_ge(self.min_capacity),
+                        total_with_option.all_ge(new_target_capacity),
+                    ]):
+                        # Only consider an option if it gets us closer to our target.
+                        if total_with_option.clamp(upper_bound=new_target_capacity).any_lt(total_target_capacity):
+                            yield (group_id, option)
+
+        def heuristic(option: Tuple[str, ClusterNodeMetadata]) -> float:
```

Comment:
As I have been thinking about this more, I am becoming increasingly uneasy about the route that we're taking here.  This change to use the heuristic is difficult to reason about, particularly for production traffic.  Note that as written, this code is going to immediately apply to all clusters/pools (including Mesos pools, Cassandra/Flink/etc).  I'm pretty worried that this change is going to cause several DARs or have other unintended consequences that will take a significant amount of time to debug.  At the very least I think we need a feature gate so that we can roll things out more slowly, and even better would be if we can split this apart into smaller chunks that are easier to reason about.